### DISCUSSION TITLE
Large memory consumption warning

### DISCUSSION THREAD

**Discussion by AlokRaj1994** [deepdoctection/deepdoctection#50] [Q&A] [answered]
UserWarning: Converting sparse IndexedSlices(IndexedSlices(indices=Tensor("tower0/gradients/tower0/cascade_rcnn_stage2/multilevel_roi_align/concat_grad/sub_2:0", shape=(?,), dtype=int64, device=/device:GPU:0), values=Tensor("tower0/gradients/tower0/cascade_rcnn_stage2/multilevel_roi_align/concat_grad/GatherV2_8:0", shape=(?, 256, 7, 7), dtype=float32, device=/device:GPU:0), dense_shape=Tensor("tower0/gradients/tower0/cascade_rcnn_stage2/multilevel_roi_align/concat_grad/Shape_2:0", shape=(4,), dtype=int32, device=/device:GPU:0))) to a dense Tensor of unknown shape. This may consume a large amount of memory.

Can you suggest something to reduce memory consumption, currently i am using  16GB RAM machine and the kernel is dead in between training. I am using 30 images only (total around 50 MB size).

**Comment by JaMe76** [comment, marked as answer]
The warning you are referring to is GPU RAM warning. According to my experience with training that model you will need at least 8 GPU RAM as the computational cost of this model is pretty expensive. 

You mentioned the kernel is dead so I take you're running the script from a notebook? In that case it is much more likely that your RAM is running OOM. It is possible to increase notebook RAM consumption of the kernel. This is one thing that might mitigate the problem.

The second thing you're referring to is the number of images. What happens under the hood when running the training script is that all datapoints from the dataflow will be cached to store annotations. When you customize the dataflow so that image have to be loaded instantly you will very quicky run OOM as images are stored as numpy images where a 1000x2000x3 will consume 45.78 MB RAM. 
However, you do not need to load images into memory. When running the training script the memory consuming part e.g. loading image, augmenting image will happen for an image only just before it gets loaded into the model and immediately gets released thereafter. 
Referring to you custom dataset in the discussion setting I suggest to set `load_image=False` in your build method. The location of the image will be cached as well and when the image is selected to enter the model for training, only then it will be loaded into memory. With that loading strategy you can easily train with 500k images, provided the annotation file gets not too large. This is basically the only memory constraint and should be the bottleneck.
